go/cmd/dolt/commands/sqlserver/server.go

```diff
@@ -278,85 +278,82 @@ func Serve(
+			GrpcListenAddr: listenaddr,
+		})
+		if err != nil {
 			lgr.Errorf("error creating SQL engine context for remotesapi server: %v", err)
 			startError = err
 			return
 		}
+
+		ctxFactory := func() (*sql.Context, error) { return sqlEngine.NewDefaultContext(ctx) }
+		authenticator := newAuthenticator(ctxFactory, sqlEngine.GetUnderlyingEngine().Analyzer.Catalog.MySQLDb)
+		args = sqle.WithUserPasswordAuth(args, authenticator)
+
+		args.TLSConfig = serverConf.TLSConfig
+		remoteSrv, err = remotesrv.NewServer(args)
+		if err != nil {
+			lgr.Errorf("error creating remotesapi server on port %d: %v", port, err)
+			startError = err
+			return
+		}
+		listeners, err := remoteSrv.Listeners()
+		if err != nil {
+			lgr.Errorf("error starting remotesapi server listeners on port %d: %v", port, err)
+			startError = err
+			return
+		} else {
+			go remoteSrv.Serve(listeners)
+		}
 	}
 
 	var clusterRemoteSrv *remotesrv.Server
 	if clusterController != nil {
-		if remoteSrvSqlCtx, err := sqlEngine.NewDefaultContext(ctx); err == nil {
-			args := clusterController.RemoteSrvServerArgs(remoteSrvSqlCtx, remotesrv.ServerArgs{
-				Logger: logrus.NewEntry(lgr),
-			})
-
-			clusterRemoteSrvTLSConfig, err := LoadClusterTLSConfig(serverConfig.ClusterConfig())
-			if err != nil {
-				lgr.Errorf("error starting remotesapi server for cluster config, could not load tls config: %v", err)
-				startError = err
-				return
-			}
-			args.TLSConfig = clusterRemoteSrvTLSConfig
-
-			clusterRemoteSrv, err = remotesrv.NewServer(args)
-			if err != nil {
-				lgr.Errorf("error creating remotesapi server on port %d: %v", *serverConfig.RemotesapiPort(), err)
-				startError = err
-				return
-			}
-			clusterController.RegisterGrpcServices(clusterRemoteSrv.GrpcServer())
+		args, err := clusterController.RemoteSrvServerArgs(sqlEngine.NewDefaultContext, remotesrv.ServerArgs{
+			Logger: logrus.NewEntry(lgr),
+		})
+		if err != nil {
+			lgr.Errorf("error creating SQL engine context for remotesapi server: %v", err)
+			startError = err
+			return
+		}
 
-			listeners, err := clusterRemoteSrv.Listeners()
-			if err != nil {
-				lgr.Errorf("error starting remotesapi server listeners for cluster config on %s: %v", clusterController.RemoteSrvListenAddr(), err)
-				startError = err
-				return
-			}
+		clusterRemoteSrvTLSConfig, err := LoadClusterTLSConfig(serverConfig.ClusterConfig())
+		if err != nil {
+			lgr.Errorf("error starting remotesapi server for cluster config, could not load tls config: %v", err)
+			startError = err
+			return
+		}
+		args.TLSConfig = clusterRemoteSrvTLSConfig
 
-			go clusterRemoteSrv.Serve(listeners)
-			go clusterController.Run()
+		clusterRemoteSrv, err = remotesrv.NewServer(args)
+		if err != nil {
+			lgr.Errorf("error creating remotesapi server on port %d: %v", *serverConfig.RemotesapiPort(), err)
+			startError = err
+			return
+		}
+		clusterController.RegisterGrpcServices(sqlEngine.NewDefaultContext, clusterRemoteSrv.GrpcServer())
```

Comment:
Wary of using NewDefaultContext. It's not really good for much of anything that context.Background isn't, beside having a different type. In particular it doesn't have a real session in it, which you need for most I/O.

Can we get away with a `context.Context` everywhere?

Reply:
I...don't think we should do that. In particular, there's a lot of SQL-y stuff that we're calling into, and we need a `*sql.Context` in various places, even when we don't have a corresponding `mysql.Conn`. I don't think the right approach is to try to move to `context.Context` everywhere we need to do something like that. Another place this comes up is event execution. In reality, our session factory and the sql engine itself has to be prepared to configure contexts and execute things on behalf of programatic interactions which are not coming from a mysql.Conn.

Reply:
I be fair, NewDefaultContext is good at things like:

* Having a logger
* Theoretically interacting nicely with session tracking, (show full processlist? kill ...? etc.), although that could be better.
* Having a session, which can track session local system variables and can allow for access to mutable global system variables, etc.

Lots of things really...